exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/rm/DistributedQueryQueue.java

```diff
@@ -0,0 +1,272 @@
+    optionManager = context.getOptionManager();
+    clusterCoordinator = context.getClusterCoordinator();
+  }
+
+  @Override
+  public void setMemoryPerNode(long memoryPerNode) {
+    this.memoryPerNode = memoryPerNode;
+    refreshConfig();
+  }
+
+  private void assignMemory() {
+
+    // Divide up memory between queues using admission rate
+    // to give more memory to larger queries and less to
+    // smaller queries. We assume that large queries are
+    // larger than small queries by a factor of
+    // largeToSmallRatio.
+
+    double totalUnits = largeToSmallRatio * largeQueueSize + smallQueueSize;
+    double memoryUnit = memoryPerNode / totalUnits;
+    memoryPerLargeQuery = Math.round(memoryUnit * largeToSmallRatio);
+    memoryPerSmallQuery = Math.round(memoryUnit);
+
+    logger.debug("Distributed queue memory config: total memory = {}, large/small memory ratio = {}",
+                 memoryPerNode, largeToSmallRatio);
+    logger.debug("Small queue: {} slots, {} bytes per slot", smallQueueSize, memoryPerSmallQuery);
+    logger.debug("Large queue: {} slots, {} bytes per slot", largeQueueSize, memoryPerLargeQuery);
+  }
+
+  @Override
+  public long getDefaultMemoryPerNode(double cost) {
+    return (cost < queueThreshold) ? memoryPerSmallQuery : memoryPerLargeQuery;
+  }
+
+  /**
+   * This limits the number of "small" and "large" queries that a Drill cluster will run
+   * simultaneously, if queuing is enabled. If the query is unable to run, this will block
+   * until it can. Beware that this is called under run(), and so will consume a Thread
+   * while it waits for the required distributed semaphore.
+   *
+   * @param queryId query identifier
+   * @param totalCost the query plan
+   * @throws QueryQueueException
+   * @throws QueueTimeoutException
+   */
+
+  @SuppressWarnings("resource")
+  @Override
+  public QueueLease enqueue(QueryId queryId, double cost) throws QueryQueueException, QueueTimeoutException {
+    final String queueName;
+    DistributedLease lease = null;
+    long queryMemory;
+    final DistributedSemaphore distributedSemaphore;
+    try {
+
+      // Only the refresh and queue computation is synchronized.
+
+      synchronized(this) {
+        refreshConfig();
+
+        // get the appropriate semaphore
+        if (cost >= queueThreshold) {
+          distributedSemaphore = clusterCoordinator.getSemaphore("query.large", largeQueueSize);
+          queueName = "large";
+          queryMemory = memoryPerLargeQuery;
+        } else {
+          distributedSemaphore = clusterCoordinator.getSemaphore("query.small", smallQueueSize);
+          queueName = "small";
+          queryMemory = memoryPerSmallQuery;
+        }
+      }
+      logger.debug("Query {} with cost {} placed into the {} queue.",
+                   QueryIdHelper.getQueryId(queryId), cost, queueName);
+
+      lease = distributedSemaphore.acquire(queueTimeout, TimeUnit.MILLISECONDS);
```

Comment:
Don't turn on queueing for high concurrency. This option should never be used for versions of Drill designed for very high query rates.

In such a high-concurrency scenario, the app using Drill might consider the in-process queue based on Java concurrency objects. This would be particularly useful if the high-concurrency case ran the query only in the Foreman Drillbit. Of course, if the goal is very high concurrency, then queueing is not very useful as its whole point is to limit concurrency (to avoid overloading memory and CPU.)

In testing, the ZK queueing overhead turned out to be (surprisingly) quite negligible for a normal, fully-distributed query. Probably the cost of one additional network request is trivial compared to the number of network hops needed to get a distributed query running across Drillbits.